Fixes TransformControl breakage.

tschw opened the pull request:
Uses "material.visible" for invisible objects that need to be used with Raycaster.

tschw committed:
Fixed TransformControls (invisible intersection).

mrdoob commented:
Uh oh! A simpler fix should be to set `material.visible` to `false` instead though.

tschw commented:
That's essentially what it is - keeping the debug facilities, enabled via `showPickers` and `showActivePlane`, intact.

The visibility of the planes is switched interactively, when `showActivePlane`. 

The materials for the "pickers" (objects for intersection tests) are setup in an initialization block and lines are already long. Adding `visible: false` all over the place might have been more code and would have been less readable.

I think my changes are sensible and I don't see a substantially simpler/clearer way without removing existing functionality. 

The debug code is now nicely wrapped in `if` blocks and can easily be spotted and can optimize out under DCE of a sufficiently smart minifier. That code helped me figure out what's wrong, so you have my vote for keeping it.

tschw commented:
Closes #6804

mrdoob commented:
> I think my changes are sensible and I don't see a substantially simpler/clearer way without removing existing functionality.

Yeah, that's the thing, maybe we can remove the debug stuff... I'll take a look.

tschw commented:
Then you might as well merge and strip it out from there:
- Eliminate all dead `if` blocks under `showActivePlane === false && showPickers === false`
- Use a single, always invisible material for all the pickers (the individual materials are currently overhead when "NDEBUG")
- Throw out the last `if` block from the diff.
- Maybe there's more that can be removed.

But it doesn't have to be *my* code that fixes it. Main thing is that it gets fixed at all ^_^.

mrdoob commented:
Yep! That sounds good! Feel free to do the changes, otherwise I'll try to do the clean up after I'm doing bringing back morphtargets+shadows... 😅

tschw committed:
Fixed broken key code for '-' in TransformControls example.

tschw committed:
Removed debug modes on mrdoob's request. + Styling.

tschw commented:
The rotation controller could use some more love. The overall feel is quirky: Axes sometimes become irresponsive in local mode, which could be a relatively young bug, and the calculation of the angle is as clumsy as it always has been.

Let's merge first. I will take a deeper look at it in a while if noone else steps up.

dubejf committed:
Use Raycaster.setFromCamera() method

Convert two uses of raycaster to using setFromCamera(). One touches
TransformControls but should not affect #6795.

Also fix an error in webgl_interactive_draggablecubes.html when the
cursor is dragged outside of the backing "plane" (which is a large
planar mesh, not an infinite plane), returning no intersections. Now,
when the limit of the bounded plane is reached, the cubes just stop
following the cursor but no error is produced.

Fix tested with webgl_interactive_draggablecubes.html and the editor.

tschw committed:
Fixes irresponsive axes.

tschw commented:
Whoops! Seems it was me who broke it :sweat_smile:. Now OK.

mrdoob commented on examples/js/controls/TransformControls.js:
```diff
@@ -161,36 +165,45 @@
 		};
 
 		this.hide = function () {
-			this.traverse(function( child ) {
```
I think this shouldn't be needed now, right? I think this code is there because, at the time, `WebGLRenderer` would render the object's children regardless of the parents `visible`.

So `this.visible = false` and `this.visible = true` should be all I think.

tschw replied:
Since visible is a public property makes me wonder why we'd want `show` and `hide` methods on the internal classes at all.

mrdoob replied:
Yeah, we can also add a warning that these are not needed anymore.

mrdoob replied:
Oh wait. Are these used internally only?

tschw replied:
See #6808, you were like a second too fast.

tschw replied:
And, yes, to answer your question. You can't easily use these Gizmo classes individually. No methods for event handlers, for instance.

tschw committed:
Changed pointerVector to Vector2.

mrdoob commented:
Thanks!

mrdoob commented:
Looking good!
Scale mode in world space seems to be broken though.

tschw commented:
I'm pretty certain that it's a limitation and never worked. In `setMode` if `mode === 'scale'` forces `space` to `'local'`.

mrdoob commented:
Indeed... It sure doesn't work on `r71`...
